### DISCUSSION TITLE
Installing New Passbands Header Issue

### DISCUSSION THREAD

**Discussion by astahl7** [phoebe-project/phoebe2#583] [General]
I'm trying to upload and fit several different lightcurves of the same object taken at different bandpasses. I have lightcurves taken in the V, R, I, J, H, and K bands, and though I see that Johnson VRI are included in PHOEBE, I can't find any of JHK listed as currently available/downloadable.

The JHK data I have was taken with AndiCAM (filter info here: https://noirlab.edu/science/programs/ctio/filters/andicam). I've tried converting the online passband information into the .fits format of the other files under /.phoebe/atmospheres/tables/passbands/, but when I import phoebe, it gives me the following error:

```
failed to load passband at /home/as179/.phoebe/atmospheres/tables/passbands/andicam_j.fits
Traceback (most recent call last):
  File "/home/as179/.conda/envs/phoebe/lib/python3.7/site-packages/phoebe/frontend/bundle.py", line 6281, in add_dataset
    ignore_none=True)
  File "/home/as179/.conda/envs/phoebe/lib/python3.7/site-packages/phoebe/parameters/parameters.py", line 3413, in set_value_all
    param.set_value(value=value, **kwargs)
  File "/home/as179/.conda/envs/phoebe/lib/python3.7/site-packages/phoebe/parameters/parameters.py", line 401, in _send_if_client
    return fctn(self, *args, **kwargs)
  File "/home/as179/.conda/envs/phoebe/lib/python3.7/site-packages/phoebe/parameters/parameters.py", line 7798, in set_value
    self._choices = list_passbands(refresh=True)
  File "/home/as179/.conda/envs/phoebe/lib/python3.7/site-packages/phoebe/atmospheres/passbands.py", line 3542, in list_passbands
    return list(set(list_installed_passbands(refresh) + list_online_passbands(refresh, repeat_errors=False)))
  File "/home/as179/.conda/envs/phoebe/lib/python3.7/site-packages/phoebe/atmospheres/passbands.py", line 3573, in list_installed_passbands
    _init_passbands(True)
  File "/home/as179/.conda/envs/phoebe/lib/python3.7/site-packages/phoebe/atmospheres/passbands.py", line 2997, in _init_passbands
    _init_passband(os.path.join(path, f))
  File "/home/as179/.conda/envs/phoebe/lib/python3.7/site-packages/phoebe/atmospheres/passbands.py", line 2946, in _init_passband
    pb = Passband.load(fullpath, load_content=False)
  File "/home/as179/.conda/envs/phoebe/lib/python3.7/site-packages/phoebe/atmospheres/passbands.py", line 433, in load
    self.ptf_order = header['ptforder']
  File "/home/as179/.conda/envs/phoebe/lib/python3.7/site-packages/astropy/io/fits/header.py", line 157, in __getitem__
    card = self._cards[self._cardindex(key)]
  File "/home/as179/.conda/envs/phoebe/lib/python3.7/site-packages/astropy/io/fits/header.py", line 1746, in _cardindex
    raise KeyError(f"Keyword {keyword!r} not found.")
KeyError: "Keyword 'PTFORDER' not found."
```

It seems like PHOEBE wants all the header information it's used to. I can insert placeholders for most since they seem to only be needed for documentation:

    primary_hdu.header['PHOEBEVN'] = '2.2.1'
    primary_hdu.header['VERSION'] = '1.0'
    primary_hdu.header['TIMESTMP'] = 'Wed Jan 22 15:53:39 2020'
    primary_hdu.header['COMMENTS'] = ''
    primary_hdu.header['REFERENC'] = 'Me'
    
But there are a few that I'm unsure about:
   
    primary_hdu.header['CALIBRTD'] = 'T'
    primary_hdu.header['PTFORDER'] =
    primary_hdu.header['PTFEAREA'] =
    primary_hdu.header['PTFPAREA'] =
    
 What do these refer to? I'm assuming T = True, but what does it mean to be calibrated? If PTF is Passband Transmission Function, what is E area versus P area? What is order? 

Any help you could provide would be much appreciated.

**Comment by aprsa** [comment]
You are correct in noting that Johnson JHK passbands are not currently included. Before I address that, though, let me go through your other questions.

The `CALIBRTD` keyword (boolean switch) tells PHOEBE whether the passband transmission table is QE-calibrated or not; if `T`, then we assume that intensity throughput is absolute, i.e. in W/m^3. If `F`, then we assume that it is relative and, in that case, intensity transformation can only be done by parametrizing via the `pblum` parameter. We try to only use calibrated PTFs, otherwise the absolute pblum mode doesn't work.

The `PTFORDER` keyword (integer) tells PHOEBE what spline order to fit to the passband transmission function in order to approximate it with an analytical function. The default (3) works for most cases.

The `PTFEAREA` and `PTFPAREA` keywords (floats) correspond to the areas under the PTF; the first is energy-weighted, and the second is photon-weighted. These values are needed to calculate luminosity-to-flux conversions properly. As they depend only on the transmission function, we calculate them ahead of time and store them in the tables.

That said, you don't ever need to fill out any of this yourself. The [phoebe/atmospheres/passband](https://github.com/phoebe-project/phoebe2/blob/master/phoebe/atmospheres/passbands.py) module has everything necessary to calculate the passband files for you. Note that we are all-but-done with rewriting the module and it will be included in one of the upcoming releases; this will greatly improve and speed up how we compute passbands. For the time being, though, this is (roughly) what you might want to do:

```
import phoebe
from astropy.units import units as u

pb = phoebe.atmospheres.passbands.Passband(
    ptf='path/to/passband.ptf',
    pbset='Johnson',
    pbname='J',
    effwl=10500,
    wlunits=u.AA,
    calibrated=True,
    reference='Origin of the PTF table',
    version=1.0,
    comments='Any additional comments'
)

pb.compute_blackbody_response()

pb.compute_ck2004_response(path='path/to/ck2004', verbose=True)
pb.compute_ck2004_intensities(path='path/to/ck2004', verbose=True)
pb.compute_ck2004_ldcoeffs()
pb.compute_ck2004_ldints()

pb.compute_phoenix_response(path='path/to/phoenix', verbose=True)
pb.compute_phoenix_intensities(path='path/to/phoenix', verbose=True)
pb.compute_phoenix_ldcoeffs()
pb.compute_phoenix_ldints()

pb.save('path/to/passband.fits')
```

Thus, to compute passbands, you will need ck2004 and phoenix atmosphere tables. I'll add the links to the website shortly and follow up here.

Finally, I just started the computation of Johnson JHK passbands; it takes ~4 hours per passband, so I should have them at some point tomorrow. I used generic PTFs but if you need an AndiCAM-specific set of PTFs, I can send you the script I used (it has a couple of extra bells compared to the script above, to impute missing Phoenix values and to add a reference to WD passband index) so that you can recreate it on your end, or I can add them to the batch if you send me the transmission functions.

**Reply by astahl7** [reply]
Fantastic! I appreciate the quick and comprehensive reply. The generic JHK PTFs should work fine -- thanks for being willing to compute them!

**Reply by aprsa** [reply]
Of course! We have woefully outdated instructions in the docs that I've been dragging my feet to fix because the new passband computation framework is almost done and it changes how this is done under the hood, hopefully for the last time. Stay tuned for that part.

**Comment by aprsa** [comment]
That took less time that I had anticipated -- Johnson JHK are now available to be used. Just please note that, while the passbands passed the initial quality control check, they have never been used so there may be problems the automated QC missed. Let me know if you run into any issues!
